src/sync/mutex.rs

```diff
@@ -94,15 +94,79 @@ impl<T: ?Sized> Mutex<T> {
                 mutex: self,
             })),
             Err(TryLockError::WouldBlock) => panic!("mutex state out of sync"),
-        }
+        };
+
+        // We need to let other threads in here so they may fail a `try_lock`
+        thread::switch();
+
+        result
     }
 
     /// Attempts to acquire this lock.
     ///
     /// If the lock could not be acquired at this time, then Err is returned. This function does not
     /// block.
     pub fn try_lock(&self) -> TryLockResult<MutexGuard<T>> {
-        unimplemented!()
+        let me = ExecutionState::me();
+
+        let mut state = self.state.borrow_mut();
+        trace!(holder=?state.holder, waiters=?state.waiters, "trying to acquire mutex {:p}", self);
+
+        if let Some(holder) = state.holder {
+            if holder == me {
+                panic!("deadlock! task {:?} tried to acquire a Mutex it already holds", me);
+            }
+        }
+
+        // We don't need a context switch here. There are two cases to analyze.
+        // * Consider that we manage to acquire the lock, but that there is another thread that
```

Comment:
Super nit: I think these comments would be easier to read if you started with `Consider that state_holder == None` and `Consider that state_holder == Some(th) and th wants to release`.  Maps more directly to the code structure, and gives a name to `the other thread`.

Reply:
+1 to basically writing this comment as a case analysis over `state.holder`